File: zuma/mopro_flutter_plugin/android/src/main/kotlin/com/example/mopro_flutter/MoproFlutterPlugin.kt
package com.example.mopro_flutter

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

import uniffi.mopro.semaphoreProve
import uniffi.mopro.semaphoreVerify
import uniffi.mopro.ProofResult

import io.flutter.plugin.common.StandardMethodCodec
/** MoproFlutterPlugin */
class MoproFlutterPlugin : FlutterPlugin, MethodCallHandler {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(
            flutterPluginBinding.binaryMessenger,
            "mopro_flutter",
            StandardMethodCodec.INSTANCE
        )
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        if (call.method == "generateProof") {
            val zkeyPath = call.argument<String>("zkeyPath") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing zkeyPath",
                null
            )

            val inputs =
                call.argument<String>("inputs") ?: return result.error(
                    "ARGUMENT_ERROR",
                    "Missing inputs",
                    null
                )

            val res = generateCircomProof(zkeyPath, inputs, ProofLib.ARKWORKS)
            val proof: ProofCalldata = toEthereumProof(res.proof)
            val convertedInputs: List<String> = toEthereumInputs(res.inputs)

            val proofList = listOf(
                mapOf(
                    "x" to proof.a.x,
                    "y" to proof.a.y
                ), mapOf(
                    "x" to proof.b.x,
                    "y" to proof.b.y
                ), mapOf(
                    "x" to proof.c.x,
                    "y" to proof.c.y
                )
            )

            // Return the proof and inputs as a map supported by the StandardMethodCodec
            val resMap = mapOf(
                "proof" to proofList,
                "inputs" to convertedInputs
            )

            result.success(
                resMap
            )
        } else if (call.method == "semaphoreProve") {
            val idSecret = call.argument<String>("idSecret") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing idSecret",
                null
            )
            val leaves = call.argument<List<String>>("leaves") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing leaves",
                null
            )
            val signal = call.argument<String>("signal") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing signal",
                null
            )
            val externalNullifier = call.argument<String>("externalNullifier") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing externalNullifier",
                null
            )

            val proofResult = semaphoreProve(idSecret, leaves, signal, externalNullifier)
            val proof = proofResult.proof
            val inputs = proofResult.inputs

            val resMap = mapOf(
                "proof" to proof,
                "inputs" to inputs
            )

            result.success(resMap)
        } else if (call.method == "semaphoreVerify") {
            val proof = call.argument<String>("proof") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing proof",
                null
            )
            val inputs = call.argument<String>("inputs") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing inputs",
                null
            )

            val proofResult = ProofResult(proof, inputs)
            val valid = semaphoreVerify(proofResult)
            result.success(valid)
        } else if (call.method == "getIdCommitment") {
            val idSecret = call.argument<String>("idSecret") ?: return result.error(
                "ARGUMENT_ERROR",
                "Missing idSecret",
                null
            )
            val commitment = getIdCommitment(idSecret)
            result.success(commitment)
        } else {
            result.notImplemented()
        }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }
}

File: zuma/mopro_flutter_plugin/ios/Classes/MoproFlutterPlugin.swift
import Flutter
import UIKit
import moproFFI

public class MoproFlutterPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "mopro_flutter", binaryMessenger: registrar.messenger())
    let instance = MoproFlutterPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    // case "generateProof":
    //     guard let args = call.arguments as? [String: Any],
    //           let zkeyPath = args["zkeyPath"] as? String,
    //           let inputs = args["inputs"] as? String
    //     else {
    //         result(FlutterError(code: "ARGUMENT_ERROR", message: "Missing arguments", details: nil))
    //         return
    //     }

    //     do {
    //         // Call the function from mopro.swift
    //         let proofResult = try generateCircomProof(zkeyPath: zkeyPath, circuitInputs: inputs, proofLib: ProofLib.arkworks)
    //         let proof = toEthereumProof(proof: proofResult.proof)
    //         let convertedInputs = toEthereumInputs(inputs: proofResult.inputs)

    //         let proofList: [[String: Any]] = [
    //             ["x": proof.a.x, "y": proof.a.y],
    //             ["x": proof.b.x, "y": proof.b.y],
    //             ["x": proof.c.x, "y": proof.c.y]
    //         ]

    //         let resMap: [String: Any] = [
    //             "proof": proofList,
    //             "inputs": convertedInputs
    //         ]

    //         // Return the proof and inputs as a map supported by the StandardMethodCodec
    //         result(resMap)
    //     } catch {
    //         result(FlutterError(code: "PROOF_GENERATION_ERROR", message: "Failed to generate proof", details: error.localizedDescription))
    //     }

    // default:
    //     result(FlutterMethodNotImplemented)
    // }
    case "semaphoreProve":
      guard let args = call.arguments as? [String: Any],
        let idSecret = args["idSecret"] as? String,
        let leaves = args["leaves"] as? [String],
        let signal = args["signal"] as? String,
        let externalNullifier = args["externalNullifier"] as? String
      else {
        result(FlutterError(code: "ARGUMENT_ERROR", message: "Missing arguments", details: nil))
        return
      }

      do {
        // Call the function from mopro.swift
        let proofResult = try semaphoreProve(
          idSecret: idSecret, leaves: leaves, signal: signal, externalNullifier: externalNullifier)

        let resMap: [String: Any] = [
          "proof": proofResult.proof,
          "inputs": proofResult.inputs,
        ]

        // Return the proof and inputs as a map supported by the StandardMethodCodec
        result(resMap)
      } catch {
        result(
          FlutterError(
            code: "PROOF_GENERATION_ERROR", message: "Failed to generate proof",
            details: error.localizedDescription))
      }

    case "semaphoreVerify":
      guard let args = call.arguments as? [String: Any],
        let proof = args["proof"] as? String,
        let inputs = args["inputs"] as? String
      else {
        result(FlutterError(code: "ARGUMENT_ERROR", message: "Missing arguments", details: nil))
        return
      }

      do {
        // Call the function from mopro.swift
        let proofResult = ProofResult(inputs: inputs, proof: proof)
        let valid = try semaphoreVerify(proof: proofResult)
        result(valid)
      } catch {
        result(
          FlutterError(
            code: "PROOF_VERIFICATION_ERROR", message: "Failed to verify proof",
            details: error.localizedDescription))
      }

    case "getIdCommitment":
      guard let args = call.arguments as? [String: Any],
        let idSecret = args["idSecret"] as? String
      else {
        result(FlutterError(code: "ARGUMENT_ERROR", message: "Missing arguments", details: nil))
        return
      }

      do {
        // Call the function from mopro.swift
        let commitment = try getIdCommitment(idSecret: idSecret)
        result(commitment)
      } catch {
        result(
          FlutterError(
            code: "COMMITMENT_GENERATION_ERROR", message: "Failed to generate commitment",
            details: error.localizedDescription))
      }

    default:
      result(FlutterMethodNotImplemented)
    }
  }
}
